python_modules/dagster/dagster/_core/instance/__init__.py

```diff
@@ -3161,6 +3164,122 @@ def add_backfill(self, partition_backfill: "PartitionBackfill") -> None:
+        # TODO(deepyaman): Abstract logic shared with `dagster-graphql`.
+        from dagster._core.definitions.partition import CachingDynamicPartitionsLoader
+        from dagster._core.definitions.selector import (
+            PartitionRangeSelector,
+            PartitionsByAssetSelector,
+            PartitionsSelector,
+        )
+        from dagster._core.execution.backfill import PartitionBackfill
+
+        backfill_id = make_new_backfill_id()
+        backfill_timestamp = get_current_timestamp()
+        backfill_datetime = datetime_from_timestamp(backfill_timestamp)
+        dynamic_partitions_store = CachingDynamicPartitionsLoader(self)
+
+        if (
+            asset_selection is not None or partition_names is not None or all_partitions
+        ) and partitions_by_assets is not None:
+            raise DagsterInvariantViolationError(
+                "partitions_by_assets cannot be used together with asset_selection or"
+                " partition_names or if all_partitions is True"
+            )
+
+        check_valid_title(title)
+
+        if asset_selection is not None:
+            backfill = PartitionBackfill.from_asset_partitions(
+                backfill_id=backfill_id,
+                asset_graph=asset_graph,
+                asset_selection=asset_selection,
+                partition_names=partition_names,
+                backfill_timestamp=backfill_timestamp,
+                tags=tags,
+                dynamic_partitions_store=dynamic_partitions_store,
+                all_partitions=all_partitions,
+                title=title,
+                description=description,
+            )
+        elif partitions_by_assets is not None:
+            partitions_by_asset_selectors = []
+            for asset_key, partitions in partitions_by_assets:
+                partitions_selector = (
+                    PartitionsSelector(ranges=[PartitionRangeSelector(*r) for r in partitions])
+                    if partitions is not None
+                    else None
+                )
+                partitions_by_asset_selectors.append(
+                    PartitionsByAssetSelector(asset_key=asset_key, partitions=partitions_selector)
+                )
+
+            backfill = PartitionBackfill.from_partitions_by_assets(
+                backfill_id=backfill_id,
+                asset_graph=asset_graph,
+                partitions_by_assets=partitions_by_asset_selectors,
+                backfill_timestamp=backfill_timestamp,
+                tags=tags,
+                dynamic_partitions_store=dynamic_partitions_store,
+                title=title,
+                description=description,
+            )
+        else:
+            raise DagsterError(
+                "Backfill requested without specifying partition set selector or asset selection"
+            )
+
+        # Check that the partition keys can be found in the asset graph.
+        # TODO(deepyaman): Abstract logic shared with `dagster-graphql`.
```

Comment:
I think this one can definitely be deduplicated; probably can be done in a subsequent PR?